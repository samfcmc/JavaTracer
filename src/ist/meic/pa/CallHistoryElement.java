package ist.meic.pa;

public class CallHistoryElement extends HistoryElement {

	public String toString() {
		return "  -> " + this.getMethodName() + " on " + this.getFileName() + ":" + this.getLineNumber();
	}	
}
