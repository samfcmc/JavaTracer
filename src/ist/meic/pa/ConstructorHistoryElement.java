package ist.meic.pa;

public class ConstructorHistoryElement extends HistoryElement {

	public String toString() {
		return "  <- ";
	}
}
