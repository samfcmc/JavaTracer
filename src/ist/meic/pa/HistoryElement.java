/**
 * 
 */
package ist.meic.pa;

import java.util.ArrayList;
import java.util.List;

/**
 * @author ist169350
 * 
 */
public abstract class HistoryElement {

	private String methodName;
	private List<String> methodParameters;
	private String fileName;

	/**
	 * 
	 */
	public HistoryElement() {
		this("", new ArrayList<String>(), "");
	}

	public HistoryElement(String methodName, List<String> methodParameters,
			String fileName) {
		super();
		this.methodName = methodName;
		this.methodParameters = methodParameters;
		this.fileName = fileName;
	}

	public String getMethodName() {
		return methodName;
	}

	public void setMethodName(String methodName) {
		this.methodName = methodName;
	}

	public List<String> getMethodParameters() {
		return methodParameters;
	}

	public void setMethodParameters(List<String> methodParameters) {
		this.methodParameters = methodParameters;
	}

	public String getFileName() {
		return fileName;
	}

	public void setFileName(String fileName) {
		this.fileName = fileName;
	}

}
