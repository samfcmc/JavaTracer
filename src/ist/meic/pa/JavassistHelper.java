package ist.meic.pa;

public class JavassistHelper {
	public static String addConstructorCallHistory(String object, String fileName, int lineNumber) {
		return "ist.meic.pa.Trace.createHistory(" + object + ");"; 
	}
}
