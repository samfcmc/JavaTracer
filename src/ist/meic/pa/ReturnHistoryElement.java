package ist.meic.pa;

public class ReturnHistoryElement extends HistoryElement {

	public String toString() {
		return "  <- " + this.getMethodName() + " on " + this.getFileName() + ":" + this.getLineNumber();
	}
}
