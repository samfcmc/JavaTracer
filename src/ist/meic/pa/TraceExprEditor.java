package ist.meic.pa;

import javassist.CannotCompileException;
import javassist.NotFoundException;
import javassist.expr.ExprEditor;
import javassist.expr.MethodCall;
import javassist.expr.NewExpr;

public class TraceExprEditor extends ExprEditor {
	public void edit(MethodCall call) throws CannotCompileException {
		traceMethodCall(call);
	}

	public void edit(NewExpr newExpr) throws CannotCompileException {
		traceNewExpr(newExpr);
	}
	
	private void traceMethodCall(MethodCall call) throws CannotCompileException {
		try {
			String template = 
					"{" +
					"	ist.meic.pa.Trace.addUsedAsArgumentElementToHistory($args, \"%s\", \"%s\", %d);" +
					"	$_ = $proceed($$);" +
					"	ist.meic.pa.Trace.addReturnElementToHistory(($w)$_,\"%s\", \"%s\", %d);" +
					"}";
			call.replace(String.format(template,
					call.getMethod().getLongName(), call.getFileName(),
					call.getLineNumber(), call.getMethod().getLongName(),
					call.getFileName(), call.getLineNumber()));
		} catch (NotFoundException e) {
			e.printStackTrace();
		}
	}

	private void traceNewExpr(NewExpr newExpr) throws CannotCompileException {
		try {
			final String template = 
					"{" +
					"	$_ = $proceed($$);" +
					"	ist.meic.pa.Trace.addReturnElementToHistory(($w)$_, \"%s\", \"%s\", %d);" +
					"}";
			newExpr.replace(String.format(template, newExpr
					.getConstructor().getLongName(), newExpr
					.getFileName(), newExpr.getLineNumber()));
		} catch (NotFoundException e) {
			e.printStackTrace();
		}
	}
}
