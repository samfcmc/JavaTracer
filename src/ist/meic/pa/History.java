package ist.meic.pa;

import ist.meic.pa.historyelements.HistoryElement;

import java.io.PrintStream;
import java.util.ArrayList;
import java.util.List;

public class History {

	private List<HistoryElement> historyElements;

	public History() {
		this(new ArrayList<HistoryElement>());
	}

	public History(List<HistoryElement> historyElements) {
		this.historyElements = historyElements;
	}

	public void addHistoryElement(HistoryElement historyElement) {
		this.historyElements.add(historyElement);
	}

	public void print(PrintStream printStream) {
		for (HistoryElement historyElement : this.historyElements) {
			printStream.println(historyElement.toString());
		}
	}

}
