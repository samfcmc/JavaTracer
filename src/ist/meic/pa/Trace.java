package ist.meic.pa;

import ist.meic.pa.historyelements.CastHistoryElement;
import ist.meic.pa.historyelements.Direction;
import ist.meic.pa.historyelements.FieldAccessHistoryElement;
import ist.meic.pa.historyelements.HistoryElement;
import ist.meic.pa.historyelements.ReturnHistoryElement;
import ist.meic.pa.historyelements.UsedAsArgumentHistoryElement;

import java.io.PrintStream;
import java.util.IdentityHashMap;
import java.util.Map;

public class Trace {

	private static Map<Object, History> history = new IdentityHashMap<Object, History>();
	private static PrintStream printStream = System.err;
	private static final String NO_HISTORY = "Tracing for %s is nonexistent!";
	private static final String OBJECT_HISTORY = "Tracing for %s";

	public static void print(Object object) {
		History objectHistory = getHistory(object);

		if (objectHistory == null) {
			String message = String.format(NO_HISTORY, object.toString());
			printStream.println(message);
		} else {
			String message = String.format(OBJECT_HISTORY, object.toString());
			printStream.println(message);
			objectHistory.print(printStream);
		}
	}

	public static History getHistory(Object object) {
		return history.get(object);
	}

	public static void createHistory(Object object) {
		if (!history.containsKey(object)) {
			history.put(object, new History());
		}
	}

	public static void addReturnElementToHistory(Object object,
			String methodName, String fileName, int lineNumber) {
		createHistory(object);
		HistoryElement element = new ReturnHistoryElement(methodName, fileName,
				lineNumber);
		getHistory(object).addHistoryElement(element);
	}

	public static void addUsedAsArgumentElementToHistory(Object[] objects,
			String methodName, String fileName, int lineNumber) {
		HistoryElement historyElement = new UsedAsArgumentHistoryElement(
				methodName, fileName, lineNumber);
		for (Object object : objects) {
			createHistory(object);
			getHistory(object).addHistoryElement(historyElement);
		}
	}

	public static void addCastElementToHistory(Object object,
			Class<?> castType, String fileName, int lineNumber) {
		HistoryElement element = new CastHistoryElement(castType.getName(),
				fileName, lineNumber);
		createHistory(object);
		getHistory(object).addHistoryElement(element);
	}

	private static void addFieldAccessElementToHistory(Object object,
			String fieldName, String fileName, int lineNumber,
			Direction direction) {
		HistoryElement element = new FieldAccessHistoryElement(fieldName,
				fileName, lineNumber, direction);
		createHistory(object);
		getHistory(object).addHistoryElement(element);
	}

	public static void addFieldSetElementToHistory(Object object,
			String fieldName, String fileName, int lineNumber) {
		addFieldAccessElementToHistory(object, fieldName, fileName, lineNumber,
				Direction.IN);
	}

	public static void addFieldGetElementToHistory(Object object,
			String fieldName, String fileName, int lineNumber) {
		addFieldAccessElementToHistory(object, fieldName, fileName, lineNumber,
				Direction.OUT);
	}
}
