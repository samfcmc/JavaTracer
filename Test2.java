import ist.meic.pa.Trace;

import java.util.*;

class Test {

    Map m = new HashMap();

    public Object identity(Object o) {
        return o;
    }

    public void test() {
        Object o = new String("MyObj");

        identity(o);
        m.put(2,o);
        m.get(2);

        Trace.print(o);

        for (Object obj : m.values()) {
            System.out.println(obj);
        }

        Trace.print(o);

    }
}

public class Test2 {
    public static void main(String args[]) {
        (new Test()).test();
    }
}