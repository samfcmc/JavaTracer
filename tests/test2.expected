Tracing for MyObj
  <- java.lang.String(java.lang.String) on Test2.java:14
  -> Test.identity(java.lang.Object) on Test2.java:16
  <- Test.identity(java.lang.Object) on Test2.java:16
  -> java.util.Map.put(java.lang.Object,java.lang.Object) on Test2.java:17
  <- java.util.Map.get(java.lang.Object) on Test2.java:18
  -> ist.meic.pa.Trace.print(java.lang.Object) on Test2.java:20
MyObj
Tracing for MyObj
  <- java.lang.String(java.lang.String) on Test2.java:14
  -> Test.identity(java.lang.Object) on Test2.java:16
  <- Test.identity(java.lang.Object) on Test2.java:16
  -> java.util.Map.put(java.lang.Object,java.lang.Object) on Test2.java:17
  <- java.util.Map.get(java.lang.Object) on Test2.java:18
  -> ist.meic.pa.Trace.print(java.lang.Object) on Test2.java:20
  <- java.util.Iterator.next() on Test2.java:22
  -> java.io.PrintStream.println(java.lang.Object) on Test2.java:23
  -> ist.meic.pa.Trace.print(java.lang.Object) on Test2.java:26
